src/kernel/stage3.c

```diff
@@ -147,9 +156,13 @@ closure_function(3, 0, void, startup,
     closure_finish();
 }
 
-thunk create_init(kernel_heaps kh, tuple root, filesystem fs)
+thunk create_init(kernel_heaps kh, tuple root, filesystem fs, merge *m)
 {
-    return closure(heap_general(kh), startup, kh, root, fs);
+    heap h = heap_general(kh);
+    status_handler completion = closure(h, program_start, 0, 0);
+    *m = allocate_merge(h, completion);
+    apply_merge(*m);
```

Comment:
This is now fixed by passing the status handler returned by apply_merge() to the startup closure (and then from there to read_program_complete). read_program_complete now applies this status_handler instead of the one that is passed to storage_when_ready().